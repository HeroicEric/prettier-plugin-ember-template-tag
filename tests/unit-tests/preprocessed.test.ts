import { getTemplateLocals } from '@glimmer/syntax';
import { Preprocessor } from 'content-tag';
import { preprocessEmbeddedTemplates } from 'ember-template-imports/lib/preprocess-embedded-templates.js';
// FIXME: Do we need this?
// import {
//   TEMPLATE_TAG_NAME,
//   TEMPLATE_TAG_PLACEHOLDER,
// } from 'ember-template-imports/lib/util.js';
import { describe, expect, test } from 'vitest';

import { AMBIGUOUS_PLACEHOLDER } from '../helpers/ambiguous.js';
import type { TestCase } from '../helpers/cases.js';
import { getAllCases } from '../helpers/cases.js';
import { format } from '../helpers/format.js';
import type { Config } from '../helpers/make-suite.js';
import { makeSuite } from '../helpers/make-suite.js';

// FIXME: Are these still accurate?
export const TEMPLATE_TAG_NAME = 'template';
export const TEMPLATE_TAG_PLACEHOLDER = '__GLIMMER_TEMPLATE';

const p = new Preprocessor();

describe('format', () => {
  makeSuite(getAllCases, preprocessedTest)({ name: 'with preprocessed code' });

  /*
   FIXME: We might not be able to maintain compatibility with
   ember-template-imports preprocessed code, which would mean we'd have to
   drop support for it altogether (and remove the set of tests below).

   This would cause compat issues with eslint-plugin-ember until they are using
   content-tag. Additionally, if they start using content-tag, running prettier
   via eslint-plugin-ember won't be possible until _this_ library supports
   content-tag.
  */
  makeSuite(
    getAllCases,
    preprocessedTestLegacy,
  )({ name: 'with legacy ember-template-imports preprocessed code' });
});

function preprocessedTest(config: Config, testCase: TestCase): void {
  test(`it formats ${testCase.name}`, async () => {
    const code = testCase.code.replaceAll(AMBIGUOUS_PLACEHOLDER, '');
    const preprocessed = p.process(code);
    const result = await format(preprocessed, config.options);
    expect(result).toMatchSnapshot();
  });
}

function preprocessedTestLegacy(config: Config, testCase: TestCase): void {
  test(`it formats ${testCase.name}`, async () => {
    const code = testCase.code.replaceAll(AMBIGUOUS_PLACEHOLDER, '');
    const preprocessed = preprocessEmbeddedTemplates(code, {
      getTemplateLocals,

      templateTag: TEMPLATE_TAG_NAME,
      templateTagReplacement: TEMPLATE_TAG_PLACEHOLDER,

      includeSourceMaps: false,
      includeTemplateTokens: false,

      relativePath: testCase.path,
    }).output;
    const result = await format(preprocessed, config.options);
    expect(result).toMatchSnapshot();
  });
}
