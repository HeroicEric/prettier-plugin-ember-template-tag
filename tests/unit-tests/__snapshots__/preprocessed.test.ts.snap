// Vitest Snapshot v1

exports[`format > config > with preprocessed code > it formats ../cases/gjs/component-class.gjs 1`] = `
"import Component from \\"@glimmer/component\\";

/** It's a component */
class MyComponent extends Component {
  [__GLIMMER_TEMPLATE(
    \`<h1>
      Class top level template. Class top level template. Class top level
      template. Class top level template. Class top level template.
    </h1>\`,
    { strictMode: true }
  )]
}
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/component-class-with-content-before-template.gjs 1`] = `
"import Component from \\"@glimmer/component\\";

/** It's a component */
class MyComponent extends Component {
  get whatever() {}

  [__GLIMMER_TEMPLATE(
    \`<h1>
      Class top level template. Class top level template. Class top level
      template. Class top level template. Class top level template.
    </h1>\`,
    { strictMode: true }
  )]
}
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/component-class-with-template-literal.gjs 1`] = `
"import Component from \\"@glimmer/component\\";

/** It's a component */
class MyComponent extends Component {
  [__GLIMMER_TEMPLATE(
    \`<h1>
      Class top level template. Class top level template. Class top level
      template. Class top level template. Class top level template.
    </h1>\`,
    { strictMode: true }
  )]

  what = \`template literal that is not a template\`;
}
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/default-export.gjs 1`] = `
"[
  __GLIMMER_TEMPLATE(
    \`Explicit default export module top level component. Explicit default export
    module top level component. Explicit default export module top level
    component. Explicit default export module top level component. Explicit
    default export module top level component.\`,
    { strictMode: true }
  ),
]
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/exported-mod-var.gjs 1`] = `
"export const Exported = [
  __GLIMMER_TEMPLATE(
    \`Exported variable template. Exported variable template. Exported variable
    template. Exported variable template. Exported variable template. Exported
    variable template. Exported variable template.\`,
    { strictMode: true }
  ),
];
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/js-only.gjs 1`] = `
"const num = 1;
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/mod-var.gjs 1`] = `
"const Private = [
  __GLIMMER_TEMPLATE(
    \`Private variable template. Private variable template. Private variable
    template. Private variable template. Private variable template. Private
    variable template. Private variable template.\`,
    { strictMode: true }
  ),
];
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/multiple-declarations.gjs 1`] = `
"const ModVar1 = [
    __GLIMMER_TEMPLATE(
      \`<h1>
        Module variable template. Module variable template. Module variable
        template. Module variable template. Module variable template. Module
        variable template. Module variable template. Module variable template.
      </h1>\`,
      { strictMode: true }
    ),
  ],
  ModVar2 = [
    __GLIMMER_TEMPLATE(\`Second module variable template.\`, {
      strictMode: true,
    }),
  ],
  num = 1;

const bool = false,
  ModVar3 = [
    __GLIMMER_TEMPLATE(
      \`<h1>
        Module variable template. Module variable template. Module variable
        template. Module variable template. Module variable template. Module
        variable template. Module variable template. Module variable template.
      </h1>\`,
      { strictMode: true }
    ),
  ],
  ModVar4 = [
    __GLIMMER_TEMPLATE(\`Second module variable template.\`, {
      strictMode: true,
    }),
  ];
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/one-line.gjs 1`] = `
"const Oneline = [
  __GLIMMER_TEMPLATE(\`Module variable template (one line).\`, {
    strictMode: true,
  }),
];
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/prettier-ignore/component-class.gjs 1`] = `
"import Component from \\"@glimmer/component\\";

/** It's a component */
class MyComponent extends Component {
  // prettier-ignore
  [__GLIMMER_TEMPLATE(\`


    <h1>   Class top level template. Class top level template. Class top level template. Class top level template. Class top level template. </h1>
  \`, { strictMode: true })]
}
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/prettier-ignore/component-class-with-template-literal.gjs 1`] = `
"import Component from \\"@glimmer/component\\";

/** It's a component */
class MyComponent extends Component {
  // prettier-ignore
  [__GLIMMER_TEMPLATE(\`


    <h1>   Class top level template. Class top level template. Class top level template. Class top level template. Class top level template. </h1>
  \`, { strictMode: true })]

  what = \`template literal that is not a template\`;
}
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/prettier-ignore/default-export.gjs 1`] = `
"// prettier-ignore
export default [__GLIMMER_TEMPLATE(\`     Explicit default export module top level component. Explicit default export module top level component. Explicit default export module top level component. Explicit default export module top level component. Explicit default export module top level component. \`, { strictMode: true })]
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/prettier-ignore/exported-mod-var.gjs 1`] = `
"// prettier-ignore
export const Exported = [__GLIMMER_TEMPLATE(\`       Exported variable template. Exported variable template.  Exported variable template.  Exported variable template.  Exported variable template. Exported variable template. Exported variable template. \`, { strictMode: true })]
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/prettier-ignore/js-only.gjs 1`] = `
"// prettier-ignore
const num = 1;
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/prettier-ignore/multiple-declarations.gjs 1`] = `
"// prettier-ignore
const ModVar1 = [__GLIMMER_TEMPLATE(\`

  <h1>   Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template.   </h1>
\`, { strictMode: true })],
ModVar2 = [__GLIMMER_TEMPLATE(\`
  Second module variable template.
\`, { strictMode: true })],
    num = 1;

// prettier-ignore
const bool = false, ModVar3 =[__GLIMMER_TEMPLATE(\`

  <h1>   Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template.   </h1>
\`, { strictMode: true })],
ModVar4 = [__GLIMMER_TEMPLATE(\`
  Second module variable template.
\`, { strictMode: true })]
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/prettier-ignore/one-line.gjs 1`] = `
"// prettier-ignore
const Oneline = [__GLIMMER_TEMPLATE(\`      Module variable template (one line). \`, { strictMode: true })]
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/prettier-ignore/simple.gjs 1`] = `
"// prettier-ignore
[__GLIMMER_TEMPLATE(\`
              what
            \`, { strictMode: true })]
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gjs/simple.gjs 1`] = `
"[__GLIMMER_TEMPLATE(\`what\`, { strictMode: true })]
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/component-class.gts 1`] = `
"import Component from \\"@glimmer/component\\";

export interface Signature {
  Element: HTMLElement;
  Args: {
    myArg: string;
  };
  Yields: [];
}

/** It's a component */
class MyComponent extends Component<Signature> {
  [__GLIMMER_TEMPLATE(
    \`<h1>
      Class top level template. Class top level template. Class top level
      template. Class top level template. Class top level template.
    </h1>\`,
    { strictMode: true }
  )]
}
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/component-class-with-template-literal.gts 1`] = `
"import Component from \\"@glimmer/component\\";

export interface Signature {
  Element: HTMLElement;
  Args: {
    myArg: string;
  };
  Yields: [];
}

/** It's a component */
class MyComponent extends Component<Signature> {
  [__GLIMMER_TEMPLATE(
    \`<h1 ...attributes>
      {{@myArg}}
      Class top level template. Class top level template. Class top level
      template. Class top level template. Class top level template.
    </h1>\`,
    { strictMode: true }
  )]

  what = \`template literal that is not a template\`;
}
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/default-export.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

[
  __GLIMMER_TEMPLATE(
    \`Explicit default export module top level component. Explicit default export
    module top level component. Explicit default export module top level
    component. Explicit default export module top level component. Explicit
    default export module top level component.\`,
    { strictMode: true }
  ),
] as TemplateOnlyComponent<Signature>
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/exported-mod-var.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

export const Exported: TemplateOnlyComponent<Signature> = [
  __GLIMMER_TEMPLATE(
    \`Exported variable template. Exported variable template. Exported variable
    template. Exported variable template. Exported variable template. Exported
    variable template. Exported variable template.\`,
    { strictMode: true }
  ),
];
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/exported-mod-var-with-as.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

export const Exported = [
  __GLIMMER_TEMPLATE(
    \`Exported variable template. Exported variable template. Exported variable
    template. Exported variable template. Exported variable template. Exported
    variable template. Exported variable template.\`,
    { strictMode: true }
  ),
] as TemplateOnlyComponent<Signature>;
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/js-only.gts 1`] = `
"const num: number = 1;
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/mod-var.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

const Private: TemplateOnlyComponent<Signature> = [
  __GLIMMER_TEMPLATE(
    \`Private variable template. Private variable template. Private variable
    template. Private variable template. Private variable template. Private
    variable template. Private variable template.\`,
    { strictMode: true }
  ),
];
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/mod-var-with-as.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

const Private = [
  __GLIMMER_TEMPLATE(
    \`Private variable template. Private variable template. Private variable
    template. Private variable template. Private variable template. Private
    variable template. Private variable template.\`,
    { strictMode: true }
  ),
] as TemplateOnlyComponent<Signature>;
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/multiple-declarations.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

const ModVar1: TemplateOnlyComponent<Signature> = [
    __GLIMMER_TEMPLATE(
      \`<h1>
        Module variable template. Module variable template. Module variable
        template. Module variable template. Module variable template. Module
        variable template. Module variable template. Module variable template.
      </h1>\`,
      { strictMode: true }
    ),
  ],
  ModVar2: TemplateOnlyComponent<Signature> = [
    __GLIMMER_TEMPLATE(\`Second module variable template.\`, {
      strictMode: true,
    }),
  ],
  num = 1;

const bool: boolean = false,
  ModVar3: TemplateOnlyComponent<Signature> = [
    __GLIMMER_TEMPLATE(
      \`<h1>
        Module variable template. Module variable template. Module variable
        template. Module variable template. Module variable template. Module
        variable template. Module variable template. Module variable template.
      </h1>\`,
      { strictMode: true }
    ),
  ],
  ModVar4: TemplateOnlyComponent<Signature> = [
    __GLIMMER_TEMPLATE(\`Second module variable template.\`, {
      strictMode: true,
    }),
  ];
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/multiple-declarations-with-as.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

const ModVar1 = [
    __GLIMMER_TEMPLATE(
      \`<h1>
        Module variable template. Module variable template. Module variable
        template. Module variable template. Module variable template. Module
        variable template. Module variable template. Module variable template.
      </h1>\`,
      { strictMode: true }
    ),
  ] as TemplateOnlyComponent<Signature>,
  ModVar2 = [
    __GLIMMER_TEMPLATE(\`Second module variable template.\`, {
      strictMode: true,
    }),
  ] as TemplateOnlyComponent<Signature>,
  num = 1;

const bool: boolean = false,
  ModVar3 = [
    __GLIMMER_TEMPLATE(
      \`<h1>
        Module variable template. Module variable template. Module variable
        template. Module variable template. Module variable template. Module
        variable template. Module variable template. Module variable template.
      </h1>\`,
      { strictMode: true }
    ),
  ] as TemplateOnlyComponent<Signature>,
  ModVar4 = [
    __GLIMMER_TEMPLATE(\`Second module variable template.\`, {
      strictMode: true,
    }),
  ] as TemplateOnlyComponent<Signature>;
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/one-line.gts 1`] = `
"import type { TOC } from \\"@ember/component/template-only\\";

export interface Sig {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

const Oneline = [
  __GLIMMER_TEMPLATE(\`One line.\`, { strictMode: true }),
] as TOC<Sig>;
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/prettier-ignore/component-class.gts 1`] = `
"import Component from \\"@glimmer/component\\";

export interface Signature {
  Element: HTMLElement;
  Args: {
    myArg: string;
  };
  Yields: [];
}

/** It's a component */
class MyComponent extends Component<Signature> {
  // prettier-ignore
  [__GLIMMER_TEMPLATE(\`


    <h1>   Class top level template. Class top level template. Class top level template. Class top level template. Class top level template. </h1>
  \`, { strictMode: true })]
}
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/prettier-ignore/component-class-with-template-literal.gts 1`] = `
"import Component from \\"@glimmer/component\\";

export interface Signature {
  Element: HTMLElement;
  Args: {
    myArg: string;
  };
  Yields: [];
}

/** It's a component */
class MyComponent extends Component<Signature> {
  // prettier-ignore
  [__GLIMMER_TEMPLATE(\`


    <h1 ...attributes>   {{@myArg}} Class top level template. Class top level template. Class top level template. Class top level template. Class top level template. </h1>
  \`, { strictMode: true })]

  what = \`template literal that is not a template\`;
}
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/prettier-ignore/default-export.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}
// prettier-ignore
export default [__GLIMMER_TEMPLATE(\`     Explicit default export module top level component. Explicit default export module top level component. Explicit default export module top level component. Explicit default export module top level component. Explicit default export module top level component. \`, { strictMode: true })] as TemplateOnlyComponent<Signature>
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/prettier-ignore/exported-mod-var.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

// prettier-ignore
export const Exported: TemplateOnlyComponent<Signature>     = [__GLIMMER_TEMPLATE(\`       Exported variable template. Exported variable template.  Exported variable template.  Exported variable template.  Exported variable template. Exported variable template. Exported variable template. \`, { strictMode: true })]
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/prettier-ignore/exported-mod-var-with-as.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

// prettier-ignore
export const Exported = [__GLIMMER_TEMPLATE(\`       Exported variable template. Exported variable template.  Exported variable template.  Exported variable template.  Exported variable template. Exported variable template. Exported variable template. \`, { strictMode: true })] as TemplateOnlyComponent<Signature>
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/prettier-ignore/js-only.gts 1`] = `
"// prettier-ignore
const num:   number = 1;
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/prettier-ignore/multiple-declarations.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

// prettier-ignore
const ModVar1: TemplateOnlyComponent<Signature> = [__GLIMMER_TEMPLATE(\`

  <h1>   Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template.   </h1>
\`, { strictMode: true })],
ModVar2 : TemplateOnlyComponent<Signature>= [__GLIMMER_TEMPLATE(\`
  Second module variable template.
\`, { strictMode: true })],
    num = 1;

// prettier-ignore
const bool: boolean = false, ModVar3: TemplateOnlyComponent<Signature>  = [__GLIMMER_TEMPLATE(\`

  <h1>   Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template.   </h1>
\`, { strictMode: true })],
ModVar4: TemplateOnlyComponent<Signature> = [__GLIMMER_TEMPLATE(\`
  Second module variable template.
\`, { strictMode: true })]
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/prettier-ignore/multiple-declarations-with-as.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

// prettier-ignore
const ModVar1 = [__GLIMMER_TEMPLATE(\`

  <h1>   Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template.   </h1>
\`, { strictMode: true })] as TemplateOnlyComponent<Signature>,
ModVar2 = [__GLIMMER_TEMPLATE(\`
  Second module variable template.
\`, { strictMode: true })] as TemplateOnlyComponent<Signature>,
    num = 1;

// prettier-ignore
const bool: boolean = false, ModVar3  = [__GLIMMER_TEMPLATE(\`

  <h1>   Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template. Module variable template.   </h1>
\`, { strictMode: true })] as TemplateOnlyComponent<Signature>,
ModVar4 = [__GLIMMER_TEMPLATE(\`
  Second module variable template.
\`, { strictMode: true })] as TemplateOnlyComponent<Signature>
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/prettier-ignore/one-line.gts 1`] = `
"import type { TOC } from \\"@ember/component/template-only\\";

export interface Sig {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

// prettier-ignore
const Oneline = [__GLIMMER_TEMPLATE(\`      Module variable template (one line). \`, { strictMode: true })] as TOC<Sig>
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/prettier-ignore/simple.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

// prettier-ignore
[__GLIMMER_TEMPLATE(\`
              what
            \`, { strictMode: true })] as TemplateOnlyComponent<Signature>
"
`;

exports[`format > config > with preprocessed code > it formats ../cases/gts/simple.gts 1`] = `
"import type { TemplateOnlyComponent } from \\"@ember/component/template-only\\";

export interface Signature {
  Element: HTMLElement;
  Args: {};
  Yields: [];
}

[
  __GLIMMER_TEMPLATE(\`what\`, { strictMode: true }),
] as TemplateOnlyComponent<Signature>
"
`;
